crates/solver/src/settlement_submission.rs

```diff
@@ -55,68 +67,66 @@ impl SolutionSubmitter {
-                    &self.contract,
-                    &account,
-                    args.submit_api.as_ref(),
-                    &eden_gas_price_estimator,
-                )?;
-                submitter.submit(settlement, params).await
-            }
-            TransactionStrategy::Flashbots(args) => {
-                if !matches!(account, Account::Offline(..)) {
-                    return Err(SubmissionError::from(anyhow!(
-                        "Submission requires offline account for signing"
-                    )));
+        let is_dry_run: bool = self
+            .transaction_strategy
+            .iter()
+            .any(|strategy| matches!(strategy, TransactionStrategy::DryRun));
+
+        if is_dry_run {
+            Ok(dry_run::log_settlement(account, &self.contract, settlement).await?)
+        } else {
+            let mut futures = self
+                .transaction_strategy
+                .iter()
+                .map(|strategy| {
+                    async {
+                        match &*strategy {
+                            TransactionStrategy::Eden(_) | TransactionStrategy::Flashbots(_) => {
+                                if !matches!(account, Account::Offline(..)) {
+                                    return Err(SubmissionError::from(anyhow!(
+                                        "Submission to private network requires offline account for signing"
+                                    )));
+                                }
+                            }
+                            TransactionStrategy::CustomNodes(_) => {}
+                            TransactionStrategy::DryRun => unreachable!(),
+                        };
+
+                        let strategy_args = strategy.strategy_args().expect("unreachable code executed");
+                        let params = SubmitterParams {
+                            target_confirm_time: self.target_confirm_time,
+                            gas_estimate,
+                            deadline: Some(Instant::now() + self.max_confirm_time),
+                            retry_interval: self.retry_interval,
+                        };
+                        let gas_price_estimator = SubmitterGasPriceEstimator {
+                            inner: self.gas_price_estimator.as_ref(),
+                            additional_tip: Some(strategy_args.additional_tip),
+                            gas_price_cap: self.gas_price_cap,
+                        };
+                        let submitter = Submitter::new(
+                            &self.contract,
+                            &account,
+                            strategy_args.submit_api.as_ref(),
+                            &gas_price_estimator,
+                        )?;
+                        submitter.submit(settlement.clone(), params).await
+                    }
+                    .boxed()
+                })
+                .collect::<Vec<_>>();
+
+            loop {
```

Comment:
This logic is quite involved (what happens if there is a dry run in the list, what happens if flashbots fails but eden succeeds, etc) and would really benefit from some unit tests imo.

Reply:
DryRun does actually messes everything up, as a completely different strategy fails to fit into the flow for other strategies.
For now I went with this, but I'm open for suggestions how to improve the arguments:
`transaction_strategy is now a vector. If it contains DryRun option or if it contains several options where DryRun is one of them, then transaction_strategy is resolved to DryRun. In other cases it is resolved to other strategies or combination of other strategies.`
Regarding unit tests, this is one of those components that is hard to cover with UTs, because we have infinite loop with deadline, also I believe it's not easy to simulate the nonce change. Maybe some e2e test would be more appropriate here?

Reply:
> I'm open for suggestions how to improve

I probably don't see all the difficulties, but I personally would maybe try to make the `SolutionSubmitter` component abstract (having a trait and multiple implementations). Then you can have "simple strategy" submitters (e.g. DryRun, Flashbots only, Eden only, etc) and more complex, composed strategies (e.g. using multiple sub strategies in a vector, using sub strategy A if condition C is met, strategy B otherwise).

Such an abstraction may allow us to implement arbitrary logic (maybe one day we can get a MEV searcher to actually write us a really smart strategy).

Maybe this is a big enough "design" exercise to brainstorm with the rest of the backend team.

> Regarding unit tests, this is one of those components that is hard to cover with UTs

This is usually just as sign that the code is currently not architected in a way that allows for unit tests. Specifically because there is an endless loop, I'm concerned about not covering this with a test. If we cannot simulate a nonce change, we should make the future that gets created for submission mockable so that in the unit test we can simulate it coming back with `Ok` and `Error` (without actually requiring an actual RPC conneciton).

I don't know if an e2e test can work here, since Flashbots/Eden are hard to simulate e2e unless we try to run it on mainnet (which would be super costly on CI).